src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+    wxWebResponse(request)
+{
+    curl_easy_setopt(GetHandle(), CURLOPT_WRITEDATA, static_cast<void*>(this));
+    curl_easy_setopt(GetHandle(), CURLOPT_HEADERDATA, static_cast<void*>(this));
+}
+
+size_t wxWebResponseCURL::WriteData(void* buffer, size_t size)
+{
+    void* buf = GetDataBuffer(size);
+    memcpy(buf, buffer, size);
+    ReportDataReceived(size);
+    return size;
+}
+
+size_t wxWebResponseCURL::AddHeaderData(const char * buffer, size_t size)
+{
+    wxString hdr(buffer, size);
+    hdr.Trim();
+
+    if ( hdr.StartsWith("HTTP/") )
+    {
+        // First line of the headers contains status text after
+        // version and status
+        m_statusText = hdr.AfterFirst(' ').AfterFirst(' ');
+        m_headers.clear();
+    }
+    else if ( !hdr.empty() )
+    {
+        wxArrayString hdrArr = wxSplit(hdr, ':');
+        wxString hdrName;
+        wxString hdrValue;
+        if ( hdrArr.size() > 0 )
+            hdrName = hdrArr[0].Trim().MakeUpper();
+        if ( hdrArr.size() > 1 )
+            hdrValue = hdrArr[1].Trim(false);
+        m_headers[hdrName] = hdrValue;
+    }
+
+    return size;
+}
+
+wxInt64 wxWebResponseCURL::GetContentLength() const
+{
+#if CURL_AT_LEAST_VERSION(7, 55, 0)
+    curl_off_t len = 0;
+    curl_easy_getinfo(GetHandle(), CURLINFO_CONTENT_LENGTH_DOWNLOAD_T, &len);
+    return len;
+#else
+    double len = 0;
+    curl_easy_getinfo(GetHandle(), CURLINFO_CONTENT_LENGTH_DOWNLOAD, &len);
+    return (wxInt64)len;
+#endif
+}
+
+wxString wxWebResponseCURL::GetURL() const
+{
+    char* urlp = NULL;
+    curl_easy_getinfo(GetHandle(), CURLINFO_EFFECTIVE_URL, &urlp);
+    return wxString(urlp);
+}
+
+wxString wxWebResponseCURL::GetHeader(const wxString& name) const
+{
+    wxWebRequestHeaderMap::const_iterator it = m_headers.find(name.Upper());
+    if ( it != m_headers.end() )
+        return it->second;
+    else
+        return wxString();
+}
+
+int wxWebResponseCURL::GetStatus() const
+{
+    long status = 0;
+    curl_easy_getinfo(GetHandle(), CURLINFO_RESPONSE_CODE, &status);
+    return status;
+}
+
+//
+// wxWebRequestCURL
+//
+
+static size_t wxCURLRead(char *buffer, size_t size, size_t nitems, void *userdata)
+{
+    if ( userdata )
+        return static_cast<wxWebRequestCURL*>(userdata)->ReadData(buffer, size * nitems);
+    else
+        return 0;
+}
+
+wxWebRequestCURL::wxWebRequestCURL(wxWebSession & session, int id, const wxString & url):
+    wxWebRequest(session, id)
+{
+    m_headers = session.GetHeaders();
+    m_headerList = NULL;
+
+    m_handle = curl_easy_init();
+    // Set error buffer to get more detailed CURL status
+    m_errorBuffer[0] = '\0';
+    curl_easy_setopt(m_handle, CURLOPT_ERRORBUFFER, m_errorBuffer);
+    // Set this request in the private pointer
+    curl_easy_setopt(m_handle, CURLOPT_PRIVATE, static_cast<void*>(this));
+    // Set URL to handle
+    curl_easy_setopt(m_handle, CURLOPT_URL, static_cast<const char*>(url.mb_str()));
```

Comment:
Sorry to harp on it, but I'm pretty sure it's wrong to use the current locale encoding here. It probably should be UTF-8 but I don't know libcurl API sufficiently well to be sure, could you please check this?

Reply:
It's even worse in this case: we need to percent-encode the URL too. Done now.